### PR TITLE
Add drag and drop mode to plots

### PR DESCRIPTION
**Pull request by sroy3**
# 1/3 main <- this <- #4942 <- #4960

Part of #4867 

Demo 

https://github.com/iterative/vscode-dvc/assets/3683420/3f51aa6f-7b91-467b-8468-930b24bf9448

I'll iterate on this PR before merging it. This is big enough as is without adding more functionality. To be added in future PRs merging back into this one:

- [ ] Style for the plot while dragging (I think the titles for custom plots aren't the right ones either, to verify)
- [x] Toggle modes while dragging (seamless with  what it used to be) (#4942 )

### CODE REVIEW

File: extension/src/plots/model/collect.ts [hunk 1]
```diff
@@ -151,10 +151,10 @@ const getCustomPlotData = (
   const updatedSpec = truncateVegaSpecTitles(spec, nbItemsPerRow, height)
 
   return {
+    content: updatedSpec,
```
Comment (sroy3) [inline extension/src/plots/model/collect.ts:154]: Making this consistent with template plots

File: webview/src/plots/components/PlotsContainer.tsx [hunk 2]
```diff
@@ -55,57 +40,41 @@ export const PlotsContainer: React.FC<PlotsContainerProps> = ({
...
-  if (menu) {
-    menuItems.unshift({
-      icon: ListFilter,
-      onClickNode: <PlotsPicker {...menu} />,
-      tooltip: 'Select Plots'
-    })
-  }
+  const menuItems: IconMenuItemProps[] = []
 
-  if (addPlotsButton) {
```
Comment (sroy3) [inline webview/src/plots/components/PlotsContainer.tsx:73]: There was never a `addPlotsButton`.

File: webview/src/plots/components/PlotsContainer.tsx [hunk 3]
```diff
@@ -55,57 +40,41 @@ export const PlotsContainer: React.FC<PlotsContainerProps> = ({
...
 
   useEffect(() => {
     window.dispatchEvent(new Event('resize'))
   }, [nbItemsPerRowOrWidth, height])
 
-  const menuItems: IconMenuItemProps[] = []
+  const changeMode = () =>
+    changeDragAndDropMode(sectionKey, dispatch, isDragAndDropMode)
 
-  if (menu) {
```
Comment (sroy3) [inline webview/src/plots/components/PlotsContainer.tsx:65]: We never passed a `menu` prop

File: webview/src/plots/components/PlotsPicker.tsx
Comment (sroy3) [inline webview/src/plots/components/PlotsPicker.tsx:1]: Was only used by the list filter [here](https://github.com/iterative/vscode-dvc/pull/4934/files#r1380305376)

File: webview/src/plots/components/customPlots/customPlotsSlice.ts [hunk 4]
```diff
@@ -36,9 +36,6 @@ export const customPlotsSlice = createSlice({
   initialState: customPlotsInitialState,
   name: 'custom',
   reducers: {
-    changeDisabledDragIds: (state, action: PayloadAction<string[]>) => {
-      state.disabledDragPlotIds = action.payload
-    },
```
Comment (sroy3) [inline webview/src/plots/components/customPlots/customPlotsSlice.ts:41]: Not necessary anymore as it was used to disable the drag and drop when using the smooth plot slider.

File: webview/src/plots/components/templatePlots/templatePlotsSlice.ts [hunk 5]
```diff
@@ -37,9 +37,6 @@ export const templatePlotsSlice = createSlice({
   initialState: templatePlotsInitialState,
   name: 'template',
   reducers: {
-    changeDisabledDragIds: (state, action: PayloadAction<string[]>) => {
```
Comment (sroy3) [inline webview/src/plots/components/templatePlots/templatePlotsSlice.ts:40]: Same as https://github.com/iterative/vscode-dvc/pull/4934/files#r1380309173

**Review by julieg18 — APPROVED** [review]
Great work so far!

File: webview/src/plots/components/App.test.tsx [hunk 6]
```diff
@@ -1849,6 +1450,497 @@ describe('App', () => {
...
+      )
+
+      const plots = screen.getAllByTestId(/summary\.json/)
+
+      dragEnter(plots[0], 'custom-plots', DragEnterDirection.LEFT)
+
+      expect(screen.getByTestId('plot_drop-target')).toBeInTheDocument()
+    })
+
+    it('should show a drop a plot at the end of the custom plots when moving a plot inside the section but not over any other plot', () => {
```
Comment (mattseddon) [inline webview/src/plots/components/App.test.tsx:1691]: [Q] `should show a drop target`?

Comment (mattseddon) [reply webview/src/plots/components/App.test.tsx:1691]: or zone to be consistent with below

File: webview/src/plots/components/util.ts [hunk 7]
```diff
@@ -1,4 +1,49 @@
...
+  isDragAndDropMode: boolean
+) => {
+  switch (sectionKey) {
+    case PlotsSection.TEMPLATE_PLOTS:
+      return dispatch(toggleTemplateDragAndDrop(!isDragAndDropMode))
+
+    case PlotsSection.CUSTOM_PLOTS:
+      return dispatch(toggleCustomDragAndDrop(!isDragAndDropMode))
+
+    default:
```
Comment (mattseddon) [inline webview/src/plots/components/util.ts:47]: [C] Seems odd there is no default. Should we throw an error here is this is called with the wrong section?

[Q] This also makes me wonder whether or not there are similar issues with drag and drop in the comparison table.

File: webview/src/plots/components/SizeSliders.tsx [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+          nbItemsPerRowOrWidth: positiveNbItems
+        })
+      )
+      resizePlots(newHeight, positiveNbItems, sectionKey)
+    },
+    [dispatch, sectionKey]
+  )
+
+  const plotHeights = Object.values(PlotHeight).filter(
+    value => typeof value !== 'string'
```
Comment (mattseddon) [inline webview/src/plots/components/SizeSliders.tsx:48]: [Q] Why is this needed?

**Review by mattseddon — APPROVED** [review]

File: webview/src/plots/components/DragAndDropGrid.tsx [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+      wrapperComponent={
+        useVirtualizedGrid
+          ? {
+              component: VirtualizedGrid as React.FC<WrapperProps>,
+              props: { nbItemsPerRow }
+            }
+          : undefined
+      }
+      parentDraggedOver={parentDraggedOver}
+      disabledDropIds={[]}
```
[the remark this answers is not in the document]
Comment (julieg18) [reply webview/src/plots/components/DragAndDropGrid.tsx:79]: Apologies, I meant remove this line like this:
```suggestion
```

If I'm understanding things correctly, `disabledDropIds` defaults to an empty array anyway.

File: webview/src/plots/components/SizeSliders.tsx (hunk 8 above)
Comment (sroy3) [reply webview/src/plots/components/SizeSliders.tsx:48]: If I remember correctly, `PlotHeight` is an enum:

```
export enum PlotHeight {
  SMALLER,
  SMALL,
  REGULAR,
  SQUARE,
  VERTICAL_NORMAL,
  VERTICAL_LARGER
}
```
which means that `Object.values(PlotHeight)` will return `[0, 1, 2, 3, 4, 5, 'SMALLER', 'SMALL',...]`. This removes the string values as they duplicate the numerical values.

File: webview/src/plots/components/App.test.tsx (hunk 6 above)
Comment (sroy3) [reply webview/src/plots/components/App.test.tsx:1691]: Moved these tests and didn't notice this. I'll update.

File: webview/src/plots/components/DragAndDropGrid.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (sroy3) [reply webview/src/plots/components/DragAndDropGrid.tsx:79]: You are right, sorry. I'll remove this.
